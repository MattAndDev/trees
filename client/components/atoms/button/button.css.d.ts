declare namespace ButtonCssNamespace {
  export interface IButtonCss {
    appear: string
    spinner: string
    wrap: string
  }
}

declare const ButtonCssModule: ButtonCssNamespace.IButtonCss & {
  /** WARNING: Only available when `css-loader` is used without `style-loader` or `mini-css-extract-plugin` */
  locals: ButtonCssNamespace.IButtonCss
}

export = ButtonCssModule
