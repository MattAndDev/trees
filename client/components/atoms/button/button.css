.wrap {
  font-size: 1rem;
  padding: 0.5rem;
  background: var(--color-primary);
  border-radius: 0.25em;
  color: var(--color-foreground);
  border: none;
  line-height: 1.5rem;
  cursor: pointer;
  outline: inherit;
}

.spinner {
  display: inline-block;
  height: 1rem;
  margin: 0 0.5rem;

  line-height: 1rem;
  vertical-align: text-top;
  animation: appear 0.1s linear;
}

.spinner::after {
  content: ' ';
  display: block;
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  border: 0.1rem solid #fff;
  border-color: #fff transparent #fff transparent;
  animation: spinner 1.2s linear infinite;
}

@keyframes spinner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

@keyframes appear {
  0% {
    width: 0;
    margin: 0;
    transform: scale(0);
  }
  100% {
    margin: 0 0.5rem;
    transform: scale(1);
    width: 1rem;
  }
}
