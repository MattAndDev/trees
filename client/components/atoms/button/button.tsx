import { h, FunctionalComponent } from 'preact'
import s from './button.css'
import cn from 'classnames'
export type ButtonProps = h.JSX.HTMLAttributes<HTMLButtonElement> & {
  isLoading?: boolean
}

export const Button: FunctionalComponent<ButtonProps> = ({
  children,
  className,
  isLoading,
  type = 'button',
  ...rest
}) => {
  return (
    <button className={cn(s.wrap, className)} type={type} {...rest}>
      {children}
      {isLoading && <span className={s.spinner} />}
    </button>
  )
}
