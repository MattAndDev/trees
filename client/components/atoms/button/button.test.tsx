import { h } from 'preact'
import { render } from '@testing-library/preact'
import { Button } from './button'

it('Renders button correctly', () => {
  const { container } = render(<Button id="btn">Oh hi!</Button>)
  const elem = container.firstChild as HTMLButtonElement
  expect(elem).toMatchSnapshot()
  expect(elem.id).toEqual('btn')
  expect(elem.type).toEqual('button')
})

it('Renders loading button by default correctly', () => {
  const { container } = render(
    <Button isLoading id="btn">
      Oh hi!
    </Button>
  )
  const elem = container.firstChild as HTMLButtonElement
  expect(elem).toMatchSnapshot()
  expect(elem.querySelector('span')?.classList.contains('spinner')).toEqual(
    true
  )
})
