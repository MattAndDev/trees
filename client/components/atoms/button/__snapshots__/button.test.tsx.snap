// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders button correctly 1`] = `
<button
  class="wrap"
  id="btn"
  type="button"
>
  Oh hi!
</button>
`;

exports[`Renders loading button by default correctly 1`] = `
<button
  class="wrap"
  id="btn"
  type="button"
>
  Oh hi!
  <span
    class="spinner"
  />
</button>
`;
